### PR TITLE
[core] Fix snapshot privilege check logic

### PR DESCRIPTION
**Pull request by yunfengzhou-hub**
### Purpose

Fix snapshot privilege check logic

### Tests

Introduced `org.apache.paimon.privilege.PrivilegedCatalogTest` to verify the changes.

### API and Format

This change does not affect API or storage format

### Documentation

This change does not introduce a new feature

### CODE REVIEW

File: paimon-core/src/main/java/org/apache/paimon/privilege/PrivilegeChecker.java [hunk 1]
```diff
@@ -24,6 +24,18 @@
 
 /** Check if current user has privilege to perform related operations. */
 public interface PrivilegeChecker extends Serializable {
+    default void assertCanSelectOrInsert(Identifier identifier) {
+        try {
+            assertCanSelect(identifier);
+        } catch (Exception e) {
```
Comment (tsreaper) [inline paimon-core/src/main/java/org/apache/paimon/privilege/PrivilegeChecker.java:30]: Catch `NoPrivilegeException` only.

File: paimon-core/src/main/java/org/apache/paimon/privilege/PrivilegeChecker.java [hunk 2]
```diff
@@ -24,6 +24,18 @@
...
+            try {
+                assertCanInsert(identifier);
+            } catch (Exception e1) {
+                throw new RuntimeException(
+                        "Need privilege to select or insert to table " + identifier, e);
```
Comment (tsreaper) [inline paimon-core/src/main/java/org/apache/paimon/privilege/PrivilegeChecker.java:35]: Add `e1` as cause. Add `e` as suppressed exception.

File: paimon-core/src/main/java/org/apache/paimon/privilege/PrivilegeChecker.java [hunk 3]
```diff
@@ -24,6 +24,25 @@
...
+            } catch (NoPrivilegeException e1) {
+                NoPrivilegeException exception =
+                        new NoPrivilegeException(
+                                e1.getUser(),
+                                e1.getObjectType(),
+                                e1.getIdentifier(),
+                                PrivilegeType.SELECT,
+                                PrivilegeType.INSERT);
+                exception.addSuppressed(e);
+                throw exception;
```
Comment (tsreaper) [inline paimon-core/src/main/java/org/apache/paimon/privilege/PrivilegeChecker.java:42]:
```suggestion
                throw new NoPrivilegeException(
                        e1.getUser(),
                        e1.getObjectType(),
                        e1.getIdentifier(),
                        PrivilegeType.SELECT,
                        PrivilegeType.INSERT);
```

Now you only catch `NoProvilegeException`, so the cause of this exception is clear. No need to add suppressed or cause.

File: paimon-core/src/test/java/org/apache/paimon/privilege/PrivilegedCatalogTest.java [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+/** Tests for {@link PrivilegedCatalog}. */
+public class PrivilegedCatalogTest extends FileSystemCatalogTest {
```
Comment (tsreaper) [inline paimon-core/src/test/java/org/apache/paimon/privilege/PrivilegedCatalogTest.java:34]: Add tests in `PrivilegeProcedureITCase`. No need to create a new test class.

Comment (yunfengzhou-hub) [reply paimon-core/src/test/java/org/apache/paimon/privilege/PrivilegedCatalogTest.java:37]: PrivilegeProcedureITCase has tests that covered `HashBucketAssignerOperator#initializeState`, where `snapshotManager()` is verified.

I discovered this bug during some direct invocation of Catalog methods, so this test to verify Catalog interface is also needed.

**Pull request merged**